### ISSUE TITLE
GP model for inputs of varying size

### ISSUE THREAD

**Issue by mgarort** [cornellius-gp/gpytorch#975]
Dear GPytorch team,

I am trying to build a model in GPyTorch for inputs of different size (for instance, in the same batch I can have a vector that is, say, 100 elements long, another that is 56 elements long, and another that is 29 elements long). This would be a generalization of having interdomain inducing points, i.e. inducing points with size different from the datapoints. However, in my case, the datapoints themselves can also be of different size.

Do you have any recommendations on how to implement a model like this? I find three main difficulties:

- From what I understand, GPyTorch kernels register and remember the input size (probably they register it when they are first called?) and they complain if subsequent inputs are of different size. I think that I could bypass this by padding the smaller inputs with 0's, making a note of the real size of each input, and passing that information to the kernel, so that all the padding is ignored.

- I am using an ApproximateGP model with VariationalStrategy. VariationalStrategy concatenates the inducing points and the inputs (in line 85, `full_inputs = torch.cat([inducing_points, x], dim=-2)`), which won't work if the inducing points and the datapoints are of different size. I guess I should write my own version of VariationalStrategy?

- Pytorch doesn't have ragged tensors, so implementing a vectorized version of the kernel will be daunting. This problem is inherent to Pytorch and not GPyTorch, but any thoughts would be greatly appreciated!

Thanks a lot in advance and looking forward to your insights.

Best,

Miguel

**Comment by gpleiss** [comment]
I don't thin that writing your own variational strategy will solve these problems. Most everything in GPyTorch is written assuming that all inputs are the same dimensionality - so most everything would have to be re-written.

The best idea that I can think of: pad the smaller inputs with some extreme value (e.g. -9000). Then write a custom kernel that appropriately drops the extreme values.

**Comment by gpleiss** [comment]
What kernel function would you use that is able to handle inputs of different size?

**Comment by mgarort** [comment]
Hi @gpleiss 

Thanks for the recommendation. We are padding our inputs and that seems to be the easiest way, so I am going to close the issue. 

The downside is that so much padding uses a lot of memory, but we have to pad anyway if we want to parallelize computations within the kernel function, since PyTorch doesn't have a ragged tensor object as of yet.

We are designing a new kernel function to handle inputs of different size, so I'll report back on the specifics once we have tested that it actually works! I haven't found much on this topic in the literature; if you guys know of existing kernel functions for multiple inputs please let me know.

Best wishes and happy holidays,

Miguel
